src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -1047,6 +1048,183 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
+        std::set<std::pair<QueryTreeNodePtr, const ConstantNode *>> equal_funcs;
+
+        /// Step 2: populate from constants, to generate new comparing pair with constant in one side
+        std::function<void(const ComparePairs &, QueryTreeNodePtr, const ConstantNode *,
+            CompareType, std::unordered_set<QueryTreeNodePtr> &)> findPairs
+            = [&](const ComparePairs & pairs, QueryTreeNodePtr current, const ConstantNode * constant,
+                CompareType type, std::unordered_set<QueryTreeNodePtr> & path)
+        {
+            if (auto it = pairs.find(current); it != pairs.end())
+            {
+                for (const auto & left : it->second)
+                {
+                    if (left.second == CompareType::equals && equal_set.contains(left.first))
+                        continue;
+
+                    if (path.contains(left.first))
+                        continue;
+
+                    equal_set.insert(left.first);
+                    path.insert(left.first);
+                    CompareType compare_type = std::min(type, left.second);
+
+                    /// Non-sense to have both sides as constant, and no repeat of equal function
+                    if (constant && !left.first->as<ConstantNode>()
+                        && (compare_type != CompareType::equals
+                            || equal_funcs.insert({left.first, constant}).second))
+                    {
+                        String compare_function_name;
+                        if (compare_type == CompareType::less)
+                            compare_function_name = "less";
+                        else if (compare_type == CompareType::greater)
+                            compare_function_name = "greater";
+                        else if (compare_type == CompareType::lessOrEquals)
+                            compare_function_name = "lessOrEquals";
+                        else if (compare_type == CompareType::greaterOrEquals)
+                            compare_function_name = "greaterOrEquals";
+                        else if (compare_type == CompareType::equals)
+                            compare_function_name = "equals";
+
+                        const auto and_node = std::make_shared<FunctionNode>(compare_function_name);
+                        and_node->getArguments().getNodes().push_back(left.first->clone());
+                        and_node->getArguments().getNodes().push_back(constant->clone());
+                        and_node->resolveAsFunction(
+                            FunctionFactory::instance().get(compare_function_name, getContext()));
+                        function_node.getArguments().getNodes().push_back(and_node);
+                    }
+
+                    findPairs(pairs, left.first,
+                        constant ? constant : current->as<ConstantNode>(), compare_type, path);
+
+                    path.erase(left.first);
```

Comment:
Well, I don't think `path.erase` is useful here, It can generate duplicated condition, like
```
a > b AND a < 3 AND a > c AND b > c
```
we get 
```
(a > b) AND (a < 3) AND (a > c) AND (b > c) AND (b < 3) AND (c < 3) AND (c < 3)
```

Why not execute `check_path.clear(); equal_set.clear();` inside the loop?

Reply:
Now use the visited set, path only considers the current path, so leading to duplication.